doc/adrs/142_add_support_for_emf_models_migration.adoc

```diff
@@ -0,0 +1,104 @@
+= ADR-142 - Add the support for EMF models migration
+
+== Context
+
+We want to be able to load any model created with a previous version of Sirius Web. 
+
+== Decision
+
+Since models are serialised/deserialised with EMF, we will use the `ExtendedMetaData` mechanism from EMF.
+
+== Consequences
+
+We will extend `ExtendedMetaData` with another class  `MigrationExtendedMetaData` that will be instantiated with the current version of the model expected by Sirius Web and a list of `IMigrationParticipant`.
+This class will also override the necessary methods from `BasicExtendedMetaData` that will allow us to migrate the model (like `getLocalElement`)
+
+One important thing to consider is that BasicExtendedMetaData methods assume that the model is stored in XMl format, since we're using JSON format, we should be really cautious about using the default methods and might have to override most of them even if we don't use them all for migrating, a BasicJSonExtendedMetaData might even be needed.
+
+[source,java]
+----
+public class MigrationExtendedMetaData extends BasicExtendedMetaData {
+    private final String version;
+    private final List<IMigrationParticipant> iMigrationParticipants;
+    public MigrationExtendedMetaData(List<IMigrationParticipant> iMigrationParticipants, String version) {
+        this.iMigrationParticipants = Objects.requireNonNull(iMigrationParticipants);
+        this.version = version;
+    }
+
+    @Override
+    public EStructuralFeature getLocalElement(EClass eClass,  String namespace, String oldName) {
+        EStructuralFeature structuralFeature = eClass.getEStructuralFeature(oldName);
+        var migrationParticipants = iMigrationParticipants.stream().
+                filter(iMigrationParticipant -> iMigrationParticipant.canHandle(this.version))
+                .sorted(Comparator.comparing(IMigrationParticipant::getMigrationParticipantVersion))
+                .toList();
+
+        for (IMigrationParticipant migrationParticipant : migrationParticipants) {
+            structuralFeature = migrationParticipant.getLocalElement(eClass, namespace, oldName);
+        }
+
+        return structuralFeature;
+    }
+}
+----
+
+In each method overridden, we call the corresponding method from `IMigrationParticipant` if it has a lesser version than the one used to instantiate the `MigrationExtendedMetaData`.
```

Comment:
You are not talking about where this version will be stored? Something like a `migration` property at the root of the JSON document would be helpful but you need to serialize it and retrieve it which will require an update to sirius-emf-json. Preferably just a simple API to add whatever we need in the JSON document as a post process for exemple (but that will require a simple ADR on the EMFJson side). Anyway, you need to store something, somewhere for view, domain and other models that indicate the last version used for the migration. We could even store the name of the migration participant used to know quickly what has been the latest migration

```
{
  "json": {
    "version": "1.0",
    "encoding": "UTF-8"
  },
  "migration": {
    "lastMigrationPerformed": "NameFeatureChangelMigrationParticipant",
    "version": "202404221458"
  },
  "ns": {},
  "content": []
}
```